### PR TITLE
Updating the 'validate' command to use the graph

### PR DESCRIPTION
**Pull request by MLainer1**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-3902)

## Description
Updating the 'validate' command to use the graph

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+def validate_unknown_content(tx: Transaction, file_paths):
+    file_paths_filter = (
+        f"WHERE content_item_from.path in {str(file_paths)}" if file_paths else ""
+    )
+    query = f"""
+        MATCH (content_item_from)-[r:{RelationshipType.USES}{{mandatorily:true}}]->(n{{not_in_repository:true}})
+        {file_paths_filter}
+        RETURN content_item_from, collect(r) as relationships, collect(n) as nodes_to
+        """
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:20]:
```suggestion
MATCH (content_item_from)-[r:{RelationshipType.USES}{{mandatorily:true}}]->(n{{not_in_repository:true}})
{file_paths_filter}
RETURN content_item_from, collect(r) as relationships, collect(n) as nodes_to"""
```

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+from demisto_sdk.commands.content_graph.common import (
+    Neo4jRelationshipResult,
+    RelationshipType,
+)
+from demisto_sdk.commands.content_graph.interface.neo4j.queries.common import run_query
+
+
+def validate_unknown_content(tx: Transaction, file_paths):
+    file_paths_filter = (
+        f"WHERE content_item_from.path in {str(file_paths)}" if file_paths else ""
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:14]:
```suggestion
        f"WHERE content_item_from.path in {file_paths}" if file_paths else ""
```
fstring does that

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+
+def validate_duplicate_display_name(
+    tx: Transaction, pack_names: list
+) -> List[Tuple[graph.Node, List[graph.Node]]]:
+    pack_names_filter = f"AND a.name in {str(pack_names)}" if pack_names else ""
+
+    query = f"""
+        MATCH (a:Pack), (b:Pack)
```
Comment (dantavori) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:110]:
```suggestion
        MATCH (a:{ContentType.Pack}), (b:{ContentType.Pack})
```

File: demisto_sdk/commands/validate/validate_manager.py [hunk 4]
```diff
@@ -375,6 +379,9 @@ def run_validation(self):
             self.use_git = True
             self.is_circle = True
             is_valid = self.run_validation_using_git()
+
+        # TODO: impl graph validations for each of the above cases
+
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validate_manager.py:384]:
```suggestion
```
it's me I know

File: demisto_sdk/commands/validate/validate_manager.py [hunk 5]
```diff
@@ -2558,7 +2593,8 @@ def create_ignored_errors_list(errors_to_check):
 
         return ignored_error_list
 
-    def add_ignored_errors_to_list(self, config, section, key, ignored_errors_list):
+    @staticmethod
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validate_manager.py:2596]: please revert this for now

File: demisto_sdk/commands/common/hook_validations/graph_validator.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        self.graph = ContentGraphInterface()
+        self.ignored_errors = {}
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        return self.graph.__exit__()
+
+    def is_valid_files(self, file_paths=[]) -> bool:
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/graph_validator.py:34]:
```suggestion
    def is_valid_graph(self, file_paths=None) -> bool:
```

**Review by dantavori — CHANGES_REQUESTED** [review]
great progress!
in addition to what we already went through together, please make sure to add/update docstrings where needed, and add debug logs.

File: demisto_sdk/__main__.py [hunk 7]
```diff
@@ -458,6 +458,14 @@ def zip_packs(**kwargs) -> int:
...
     type=click.Path(resolve_path=True),
 )
+@click.option(
+    "-vg",
+    "--validate-graph",
+    is_flag=True,
+    default=False,
+    show_default=True,
+    help="Perform validations on content graph.",
+)
```
Comment (dorschw) [inline demisto_sdk/__main__.py:468]: `--validate-graph` suggests we validate the graph itself - is this what we do here?

File: demisto_sdk/__main__.py [hunk 8]
```diff
@@ -608,6 +616,7 @@ def validate(config, **kwargs):
             file_path=file_path,
             validate_all=kwargs.get("validate_all"),
             validate_id_set=kwargs["id_set"],
+            validate_graph=kwargs.get("validate_graph"),
```
Comment (dorschw) [inline demisto_sdk/__main__.py:619]: please rename here as well

File: demisto_sdk/commands/common/errors.py [hunk 9]
```diff
@@ -1748,6 +1748,37 @@
         "ui_applicable": False,
         "related_field": "",
     },
+    # GR - Graph validations
+    "uses_with_invalid_marketplaces": {
```
Comment (dorschw) [inline demisto_sdk/commands/common/errors.py:1752]: `with`?

File: demisto_sdk/commands/common/errors.py [hunk 10]
```diff
@@ -1748,6 +1748,37 @@
...
+        "ui_applicable": False,
+        "related_field": "",
+    },
+    "uses_with_invalid_toversions": {
+        "code": "GR104",
+        "ui_applicable": False,
+        "related_field": "",
+    },
+    "Invalid_marketplaces": {
+        "code": "GR105",
```
Comment (dorschw) [inline demisto_sdk/commands/common/errors.py:1778]: how is this one different from `GR100`?

File: demisto_sdk/commands/common/hook_validations/graph_validator.py (hunk 6 above)
Comment (dorschw) [reply demisto_sdk/commands/common/hook_validations/graph_validator.py:36]: add type hints

File: demisto_sdk/commands/common/hook_validations/graph_validator.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+            self.are_marketplaces_relationships_paths_valid(file_paths),
+            self.are_fromversion_relationships_paths_valid(file_paths),
+            self.are_toversion_relationships_paths_valid(file_paths),
+            self.is_file_using_unknown_content(file_paths),
+            self.is_file_display_name_already_exists(file_paths),
+            # self.is_valid_pack_metadata_marketplaces(file_paths),
+        )
+        return all(is_valid)
+
+    def validate_dependencies(self, pack_name):
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/graph_validator.py:47]: missing type hints

File: demisto_sdk/commands/common/hook_validations/graph_validator.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+        is_valid = True
+        pack_names = []
+        if file_paths:
+            pack_names = get_pack_display_name_from_file_in_pack(file_paths)
+        query_results = self.graph.get_duplicate_pack_display_name(pack_names)
+
+        if query_results:
+            for query_result in query_results:
+                content_id = query_result[0]
+                duplicate_names_id = query_result[1]
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/graph_validator.py:256]:
```suggestion
            for content_id, duplicate_names_id in query_results:
```
assuming these are all two-tuples

File: demisto_sdk/commands/common/hook_validations/graph_validator.py [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+    @error_codes("GR105")
+    def is_valid_pack_metadata_marketplaces(self, file_paths):
+        is_valid = True
+
+        query_results = self.graph.get_item_marketplaces(file_paths)
+
+        for query_result in query_results:
+            content_id = query_result[0]
+            marketplaces = query_result[1]
+            invalid_marketplaces = set(marketplaces) - set(MarketplaceVersions)
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/graph_validator.py:274]:
```suggestion
            invalid_marketplaces = set(marketplaces).difference(MarketplaceVersions)
```

File: demisto_sdk/commands/common/tools.py [hunk 14]
```diff
@@ -109,6 +109,9 @@
 colorama.init()  # initialize color palette
 
 
+FILE_SUFFIX_TO_EXTRACT = ["yml", "py", "json"]
```
Comment (dorschw) [inline demisto_sdk/commands/common/tools.py:112]: constant name is unclear, please elaborate - why only these two? where is it used?

File: demisto_sdk/commands/common/tools.py [hunk 15]
```diff
@@ -308,6 +311,26 @@ def get_files_in_dir(
...
 
 
+def get_all_content_items_files_in_dir(project_dir_list: Iterable):
+    """
+    Gets the project directory and returns the path of all yml, json and py files in it
+    Args:
+        project_dir_list: List or set with str paths
+    :return: list of content files in the current dir with str relative paths
+    """
+    files = []
```
[the remark this answers is not in the document]
Comment (dorschw) [reply demisto_sdk/commands/common/tools.py:321]: I am not a fan of double-list-comprehension, it's hard to read. Consider using `itertools.chain`

File: demisto_sdk/commands/common/tools.py [hunk 16]
```diff
@@ -308,6 +311,26 @@ def get_files_in_dir(
...
+    output = []
+    for file_path in project_dir_list:
+        files.extend(
+            get_files_in_dir(file_path, FILE_SUFFIX_TO_EXTRACT, ignore_test_files=True)
+        )
+
+    for file in files:
+        output.append(get_path_in_content_repo(file))
+
+    return output
```
Comment (dorschw) [inline demisto_sdk/commands/common/tools.py:331]:
```suggestion
    output = []
    for file_path in project_dir_list:
        files.extend(
            get_files_in_dir(file_path, FILE_SUFFIX_TO_EXTRACT, ignore_test_files=True)
        )

    for file in files:
        output.append(get_path_in_content_repo(file))

    return output
```
```suggestion
    for file_path in project_dir_list:
        files.extend(
            get_files_in_dir(file_path, FILE_SUFFIX_TO_EXTRACT, ignore_test_files=True)
        )
    return [get_path_in_content_repo(file) for file in files]

```

File: demisto_sdk/commands/common/tools.py [hunk 17]
```diff
@@ -3385,3 +3408,31 @@ def field_to_cli_name(field_name: str) -> str:
...
+
+    Args:
+        file_path (str): the absolute file path
+    """
+
+    path = str(file_path)
+    if path.startswith("Packs"):
+        return file_path
+    file_parts = path.split("Packs")
+    return f"Packs{file_parts[1]}"
```
Comment (dorschw) [inline demisto_sdk/commands/common/tools.py:3424]: what?

File: demisto_sdk/commands/common/tools.py [hunk 18]
```diff
@@ -3385,3 +3408,31 @@ def field_to_cli_name(field_name: str) -> str:
         field_name (str): the incident/indicator field name.
     """
     return re.sub(NON_LETTERS_OR_NUMBERS_PATTERN, "", field_name).lower()
+
+
+def get_path_in_content_repo(file_path):
```
Comment (dorschw) [inline demisto_sdk/commands/common/tools.py:3413]: missing type hint

Comment (dorschw) [inline demisto_sdk/commands/common/tools.py:3413]: please use pathlib instead

File: demisto_sdk/commands/common/tools.py [hunk 19]
```diff
@@ -3385,3 +3408,31 @@ def field_to_cli_name(field_name: str) -> str:
...
+
+    Args:
+        file_path (_type_): _description_
+    """
+    display_names = set()
+    for file_path in file_paths:
+        metadata = get_pack_metadata(file_path)
+        display_names.add(metadata.get("name"))
+
+    return list(display_names)
```
Comment (dorschw) [inline demisto_sdk/commands/common/tools.py:3438]:
```suggestion
    return sortred(display_names)
```
for a deterministic result

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+        for item in run_query(tx, query)
+    }
+
+
+def validate_toversion(tx: Transaction, file_paths, to_version):
+    file_paths_filter = (
+        f"AND content_item_from.path in {str(file_paths)}" if file_paths else ""
+    )
+    from_version_filter = (
+        "AND toIntegerList(split(content_item_from.toversion, '.')) >= toIntegerList([6,5,0])"
```
Comment (dorschw) [inline demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:63]: Prefer using a constant of the version instead over a hardcoded string inside a query

File: demisto_sdk/commands/common/tools.py (hunk 18 above)
Comment (dorschw) [reply demisto_sdk/commands/common/tools.py:3413]: p.s. don't we have such a function already?

Comment (dantavori) [reply demisto_sdk/commands/common/tools.py:3413]: We have `get_relative_path_from_packs_dir()` which is not using pathlib as well

Comment (dorschw) [reply demisto_sdk/commands/common/tools.py:3413]: please make it use pathlib and use it, you can pretty much copy the code from here (note the exception class is customized there, so we need to either return _something_ (`None`?) or raise a separate exception - same as `get_relative_path_from_packs_dir`

File: demisto_sdk/__main__.py (hunk 7 above)
Comment (dorschw) [reply demisto_sdk/__main__.py:468]: I suggest changing it to `--graph`, `--use-graph` or `--use-content-graph` (a bit long), as we validate the _content_ using the graph

File: demisto_sdk/commands/common/tools.py (hunk 14 above)
Comment (dorschw) [reply demisto_sdk/commands/common/tools.py:112]: then `GRAPH_SUPPORTED_FILE_TYPES` or `GRAPH_SUPPORTED_CONTENT_FILE_TYPES` would be clearer imo

File: demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+        for item in run_query(tx, query)
+    }
+
+
+def validate_fromversion(tx: Transaction, file_paths, from_version):
+    file_paths_filter = (
+        f"AND content_item_from.path in {str(file_paths)}" if file_paths else ""
+    )
+    from_version_filter = (
+        "AND toIntegerList(split(content_item_from.fromversion, '.')) >= toIntegerList([6,5,0])"
```
[the remark this answers is not in the document]
Comment (MLainer1) [reply demisto_sdk/commands/content_graph/interface/neo4j/queries/validations.py:36]: This works only for the filters, as I need the entity from_version (and others) from the graph itself

File: demisto_sdk/commands/common/tools.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (MLainer1) [reply demisto_sdk/commands/common/tools.py:321]: I like it

File: demisto_sdk/commands/common/errors.py [hunk 22]
```diff
@@ -1748,6 +1748,37 @@
         "ui_applicable": False,
         "related_field": "",
     },
+    # GR - Graph validations
+    "uses_with_invalid_marketplaces": {
+        "code": "GR100",
+        "ui_applicable": False,
+        "related_field": "",
+    },
+    "uses_with_invalid_fromversions": {
```
[the remark this answers is not in the document]
Comment (MLainer1) [reply demisto_sdk/commands/common/errors.py:1757]: I prefer using the word `invalid` as it encapsulates better the error imo

File: demisto_sdk/commands/common/errors.py [hunk 23]
```diff
@@ -1748,6 +1748,37 @@
...
+        "code": "GR102",
+        "ui_applicable": False,
+        "related_field": "",
+    },
+    "duplicate_display_name": {
+        "code": "GR103",
+        "ui_applicable": False,
+        "related_field": "",
+    },
+    "uses_with_invalid_toversions": {
```
[the remark this answers is not in the document]
Comment (MLainer1) [reply demisto_sdk/commands/common/errors.py:1772]: Same here as above

File: demisto_sdk/commands/common/tools.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (MLainer1) [reply demisto_sdk/commands/common/tools.py:321]: Well doesn't really work due to the function returning a list, will have to make in to 2 separate lists

File: demisto_sdk/commands/common/errors.py [hunk 24]
```diff
@@ -4345,3 +4371,43 @@ def nativeimage_exist_in_script_yml(script_id):
...
         )
+
+    @staticmethod
+    @error_code_decorator
+    def uses_items_not_in_marketplaces(
+        content_name: str, marketplaces: list, content_items: list
+    ):
+        return (
+            f"Content item '{content_name}' with marketplaces: '{', '.join(marketplaces)}' is using content items: "
+            f"'{', '.join(content_items)}' that have incompatible marketplaces."
```
Comment (dorschw) [inline demisto_sdk/commands/common/errors.py:4382]: content item .. can be used in the .. marketplaces, however it uses ... which is not supported in the ... marketplace

File: demisto_sdk/commands/common/errors.py [hunk 25]
```diff
@@ -4345,3 +4371,43 @@ def nativeimage_exist_in_script_yml(script_id):
...
+        )
+
+    @staticmethod
+    @error_code_decorator
+    def uses_items_with_invalid_fromversions(
+        content_name: str, fromversion: str, content_items: list
+    ):
+        return (
+            f"Content item '{content_name}' with from_version: '{fromversion}' is using content items: "
+            f"'{', '.join(content_items)}' that have incompatible from_version."
```
Comment (dorschw) [inline demisto_sdk/commands/common/errors.py:4392]: Conetnt item .. whose from_version is .. uses the content items ... , whose from_version is higher (must be equal to, or less than ..)

File: demisto_sdk/commands/common/errors.py [hunk 26]
```diff
@@ -4345,3 +4371,43 @@ def nativeimage_exist_in_script_yml(script_id):
...
+    ):
+        return (
+            f"Content item '{content_name}' with to_version: '{toversion}' is using content items: "
+            f"'{', '.join(content_items)}' that have incompatible to_version."
+        )
+
+    @staticmethod
+    @error_code_decorator
+    def using_unknown_content(content_name, unknown_content_names):
+        return f"Content item '{content_name}' is using unknown content items: {', '.join(unknown_content_names)} "
```
Comment (dorschw) [inline demisto_sdk/commands/common/errors.py:4408]: content item .. uses a content item .. which cannot be found in the repository.

File: demisto_sdk/__main__.py (hunk 7 above)
Comment (MLainer1) [reply demisto_sdk/__main__.py:468]: Decided to use `-gr` and `--graph` to simplify things

File: demisto_sdk/commands/common/hook_validations/graph_validator.py [hunk 27]
```diff
@@ -0,0 +1,255 @@
...
+        paths_with_invalid_marketplaces = (
+            self.graph.find_uses_paths_with_invalid_marketplaces(self.file_paths)
+        )
+
+        for query_result in paths_with_invalid_marketplaces:
+            content_name = query_result.name
+            relationship_data = query_result.uses
+            marketplaces = query_result.marketplaces
+            content_items = [
+                relationship.target.name
```
Comment (ilaner) [inline demisto_sdk/commands/common/hook_validations/graph_validator.py:193]: Replace target with `content_item`

**Review by ilaner — DISMISSED** [review]
Good job, please see the comments.
Please replace all occurrences of the `target` attribute to `content_item`. It supposed to be equivalent, and we plan to remove this attribute in the near future.

File: demisto_sdk/commands/common/hook_validations/graph_validator.py [hunk 28]
```diff
@@ -0,0 +1,255 @@
...
+        error_message, error_code = Errors.uses_items_with_invalid_fromversions(
+            content_name, fromversion, content_items
+        )
+        if self.handle_error(error_message, error_code, file_path, warning=warning):
+            is_valid = False
+
+        return is_valid
+
+    @error_codes("GR104")
+    def are_toversion_relationships_paths_valid(self, file_paths=None):
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/graph_validator.py:138]: apply suggestions similarly to fromversion validation

File: demisto_sdk/commands/common/hook_validations/pack_unique_files.py [hunk 29]
```diff
@@ -1035,6 +1039,15 @@ def are_valid_files(self, id_set_validations) -> str:
...
     def validate_pack_dependencies(self):
+
+        if self.validate_graph:
+            click.secho(
+                f"\nRunning pack dependencies validation on {self.pack} with the graph\n",
+                fg="bright_cyan",
+            )
+            with GraphValidator(self.specific_validations) as graph_validator:
+                graph_validator.validate_dependencies(self.pack)
+
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/pack_unique_files.py:1050]:
```suggestion
            with GraphValidator(self.specific_validations) as graph_validator:
                graph_validator.validate_dependencies(self.pack)
        else:
```
if we do graph validation, don't do id set validation.

Comment (dantavori) [reply demisto_sdk/commands/common/hook_validations/pack_unique_files.py:1050]: also I think better to do all graph validations in one place (under is_valid_content_graph), and here do the ID set validation only if `not self.validate_graph`

**Review by dantavori — CHANGES_REQUESTED** [review]
we're almost there, most notes are really small. let's work together when you come back to resolve the notes fast

File: demisto_sdk/commands/common/hook_validations/graph_validator.py [hunk 30]
```diff
@@ -0,0 +1,255 @@
...
+                relationship.target.name
+                if relationship.target.name
+                else relationship.target.object_id
+                for relationship in relationship_data
+            ]
+            file_path = query_result.path
+            error_message, error_code = Errors.uses_items_not_in_marketplaces(
+                content_name, marketplaces, content_items
+            )
+            if self.handle_error(error_message, error_code, file_path, warning=True):
```
Comment (dantavori) [inline demisto_sdk/commands/common/hook_validations/graph_validator.py:202]:
```suggestion
            if self.handle_error(error_message, error_code, file_path):
```

**Review by dantavori — APPROVED** [review]
@ilaner let me know if you still have any notes about the excludedDependencies changes

**Pull request merged**